### PR TITLE
Auto-magic FSharp.Core package selection

### PR DESCRIPTION
**Pull request by KevinRansom**
When a developer builds a new project, the build is supposed to use the version of FSharp.Core that was released along-side the build tools.

In the past we accomplished this by doing a manual edit of the FSharp.Core version, prior to the final insertion into VS.  In the distant past we specified a wild card packagereference we moved away from the wild card reference because it makes the build perform more slowly by having a network hop to each of the sources in the nuget.config file.

Anyway, we have improved the build automation and code flow such that there is no opportunity for us to manual edit the final insertion into VS, and so we need a mechanism that allows preview builds to work using the released VS, and rtm builds of our tooling to use the version of the package that is published at RTM.

So ... this is the solution:
We use an sdk flag $(_NETCoreSdkIsPreview) to detect whether the sdk tooling is a preview release, if it's a pre-release sdk tooling then by default it uses the previously shipped fsharp.core version.  If it's an rtm sdk it uses the RTM version of fsharp.core that will shipped to nuget when the RTM sdk is shipped.

It should be noted that, the implicit package resolution behaviour can be turned off in the project using: 
<DisableImplicitFSharpCoreReference>true</DisableImplicitFSharpCoreReference>

### CODE REVIEW

File: tests/fsharp/SDKTests/tests/DefaultFSharpCore - IsPreviewSdkEnablePreviewCheck.proj [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  </PropertyGroup>
+
+  <Import Project="Test.props" />
+
+  <PropertyGroup>
+    <ExpectsValueTuple>false</ExpectsValueTuple>
+    <DisableImplicitSystemValueTupleReference>true</DisableImplicitSystemValueTupleReference>
+    <DisableFSharpCorePreviewCheck>false</DisableFSharpCorePreviewCheck>
+    <ExpectsFSharpCore>true</ExpectsFSharpCore>
+    <ExpectedFSharpCorePackageVersion>$(FSharpCoreShippedPackageVersion)</ExpectedFSharpCorePackageVersion>
```
Comment (brettfo) [inline tests/fsharp/SDKTests/tests/DefaultFSharpCore - IsPreviewSdkEnablePreviewCheck.proj:16]: Is preview = true and enable preview check = true, shouldn't this be $(FSCorePackageVersion)?  E.g., the not-yet-shipped one?

File: tests/fsharp/SDKTests/tests/DefaultFSharpCore - IsNotPreviewSdkEnablePreviewCheck.proj [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  </PropertyGroup>
+
+  <Import Project="Test.props" />
+
+  <PropertyGroup>
+    <ExpectsValueTuple>false</ExpectsValueTuple>
+    <DisableImplicitSystemValueTupleReference>true</DisableImplicitSystemValueTupleReference>
+    <DisableFSharpCorePreviewCheck>false</DisableFSharpCorePreviewCheck>
+    <ExpectsFSharpCore>true</ExpectsFSharpCore>
+    <ExpectedFSharpCorePackageVersion>$(FSCorePackageVersion)</ExpectedFSharpCorePackageVersion>
```
Comment (brettfo) [inline tests/fsharp/SDKTests/tests/DefaultFSharpCore - IsNotPreviewSdkEnablePreviewCheck.proj:16]: The test comment states that it should select the already shipped version, isn't that named $(FSharpCoreShippedPackageVersion)?

File: tests/fsharp/SDKTests/tests/DefaultFSharpCore - IsNotPreviewSdkDisablePreviewCheck.proj [hunk 3]
```diff
@@ -0,0 +1,21 @@
+<Project ToolsVersion="4.0" DefaultTargets="Test" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+  <!-- Verifies that in an RTM sdk with the previewcheck disabled it selects the already RTM fsharp.core package -->
```
Comment (brettfo) [inline tests/fsharp/SDKTests/tests/DefaultFSharpCore - IsNotPreviewSdkDisablePreviewCheck.proj:3]: Comment says RTM sdk, but the property 2 lines down indicates it's a preview SDK.

File: tests/fsharp/SDKTests/tests/DefaultFSharpCore - IsNotPreviewSdk.proj [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<Project ToolsVersion="4.0" DefaultTargets="Test" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+  <!-- Verifies that in a non preview sdk it selects the recently built fsharp.core package -->
```
Comment (brettfo) [inline tests/fsharp/SDKTests/tests/DefaultFSharpCore - IsNotPreviewSdk.proj:3]: I thought a non-preview sdk should use the boring, shipped version, not the just built one?

File: src/fsharp/FSharp.Build/Microsoft.FSharp.NetSdk.props [hunk 5]
```diff
@@ -74,7 +74,12 @@ WARNING:  DO NOT MODIFY this file unless you are knowledgeable about MSBuild and
 
   <PropertyGroup>
     <DefaultValueTuplePackageVersion>4.4.0</DefaultValueTuplePackageVersion>
-    <DefaultFSharpCorePackageVersion>{{FSharpCoreShippedPackageVersion}}</DefaultFSharpCorePackageVersion>
+    <!--
+        if $(DisableFSharpCorePreviewCheck) then use existing behaviour
+        otherwise pick fsharp.core version based on $(_NETCoreSdkIsPreview)
```
Comment (brettfo) [inline src/fsharp/FSharp.Build/Microsoft.FSharp.NetSdk.props:79]: If preview, we should use version `$(FSharpCorePreviewPackageVersion)` which will look like `4.7.2-beta.*`.

**Comment by cartermp** [comment]
@KevinRansom are you still interested in taking this?

**Comment by abelbraaksma** [comment]
I think this is the follow up (breadcrumbs): https://github.com/dotnet/fsharp/pull/9845

**Pull request closed without merging**